File: Dynamic Programming/1964. Find the Longest Valid Obstacle Course at Each Position.py
from sortedcontainers import SortedList


class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        # isnt this just longest increasing subsequence?
        n = len(obstacles)
        dp = []
        res = []
        for num in obstacles:
            # print(num, dp)
            i = bisect_right(dp, num)
            # print(i, num)
            if i == len(dp):
                dp.append(num)
                res.append(i + 1)
            else:  # that means the number is smaller
                res.append(i + 1)
                dp[i] = num
            # print(dp, res)
        return res

5 0
1 0
5 1
5 2
1 1
3 2
4 3
5 4
1 2
4 4

File: Dynamic Programming/1964. Find the Longest Valid Obstacle Course at Each Position.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        vector<int> dp;
        int n = obstacles.size();
        vector<int> res;
        for (int len: obstacles) {
            // binary search to find the right most index to inser this number

            if (dp.size() == 0) {
                dp.push_back(len);
                res.push_back(1);
            } else {
                int m = dp.size();
                int l = 0;
                int r = m - 1;
                int res1 = -1;
                while (l <= r) {
                    int m = l + (r - l)/2 ;
                    if (dp[m] <= len) { // valid index
                        l = m + 1;
                        res1 = m;
                    } else {
                        r = m - 1;
                    }

                }
                res1++;
                // cout << res1 << " " << len << "\n";
                if (res1 == dp.size()) {
                    dp.push_back(len);
                } else {
                    dp[res1] = len;
                }
                res.push_back(res1 + 1);

            }

        }

        return res;

    }
};